File: ulyp-ui/src/main/java/com/ulyp/ui/PrimaryViewController.java
package com.ulyp.ui;

import com.ulyp.transport.TCallRecordLogUploadRequest;
import com.ulyp.ui.code.SourceCodeView;
import com.ulyp.ui.font.FontSizeChanger;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.event.Event;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;
import org.springframework.beans.factory.annotation.Autowired;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.net.URL;
import java.util.ResourceBundle;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Supplier;

public class PrimaryViewController implements Initializable {

    public PrimaryViewController() {
    }

    @FXML
    public VBox primaryPane;
    @FXML
    public AnchorPane processTabAnchorPane;
    @FXML
    public AnchorPane sourceCodeViewAnchorPane;
    @Autowired
    public SourceCodeView sourceCodeView;
    @Autowired
    public ProcessTabPane processTabPane;
    @Autowired
    public FontSizeChanger fontSizeChanger;

    private AggregationStrategy aggregationStrategy = new ByRecordingIdAggregationStrategy();

    private final ExecutorService uploaderExecutorService = Executors.newFixedThreadPool(1);
    Supplier<File> fileChooser;

    @Override
    public void initialize(URL url, ResourceBundle rb) {
        this.processTabAnchorPane.getChildren().add(processTabPane);

        AnchorPane.setTopAnchor(processTabPane, 0.0);
        AnchorPane.setBottomAnchor(processTabPane, 0.0);
        AnchorPane.setRightAnchor(processTabPane, 0.0);
        AnchorPane.setLeftAnchor(processTabPane, 0.0);

        this.sourceCodeViewAnchorPane.getChildren().add(sourceCodeView);

        AnchorPane.setTopAnchor(sourceCodeView, 0.0);
        AnchorPane.setBottomAnchor(sourceCodeView, 0.0);
        AnchorPane.setRightAnchor(sourceCodeView, 0.0);
        AnchorPane.setLeftAnchor(sourceCodeView, 0.0);
    }

    public void clearAll(Event event) {
        processTabPane.clear();
    }

    public void changeAggregation(Event event) {
        aggregationStrategy = new ByThreadIdAggregationStrategy();
        // TODO maybe popup
    }

    public void openRecordedDump(ActionEvent actionEvent) {
        // Without those calls font style won't be applied until user changes font for the first time
        fontSizeChanger.upscale(primaryPane.getScene());
        fontSizeChanger.downscale(primaryPane.getScene());

        final File file = fileChooser.get();

        uploaderExecutorService.submit(
                () -> {
                    if (file != null) {
                        try (InputStream inputStream = new BufferedInputStream(new FileInputStream(file))) {

                            while (inputStream.available() > 0) {
                                TCallRecordLogUploadRequest request = TCallRecordLogUploadRequest.parseDelimitedFrom(inputStream);
                                CallRecordTreeChunk chunk = new CallRecordTreeChunk(request);
                                FileRecordingsTab fileRecordingsTab = processTabPane.getOrCreateProcessTab(new FileRecordingsTabName(file, chunk.getProcessInfo()));
                                CallRecordTreeTab recordingTab = fileRecordingsTab.getOrCreateRecordingTab(aggregationStrategy, chunk);
                                recordingTab.uploadChunk(chunk);
                                Platform.runLater(recordingTab::refreshTreeView);
                            }

                        } catch (Exception e) {
                            // TODO show error dialog
                            e.printStackTrace();
                        }
                    }
                }
        );
    }
}

File: ulyp-ui/src/main/kotlin/com/ulyp/ui/PrimaryViewController.kt
package com.ulyp.ui

import com.ulyp.storage.impl.AsyncFileStorageReader
import com.ulyp.ui.code.SourceCodeView
import com.ulyp.ui.looknfeel.FontSizeChanger
import com.ulyp.ui.looknfeel.Theme
import com.ulyp.ui.looknfeel.ThemeManager
import com.ulyp.ui.util.CssClass
import com.ulyp.ui.util.StyledText
import javafx.application.Platform
import javafx.event.ActionEvent
import javafx.event.Event
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.geometry.Insets
import javafx.scene.Scene
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.scene.text.Text
import javafx.stage.Modality
import javafx.stage.Stage
import java.io.File
import java.net.URL
import java.util.*
import java.util.function.Supplier


class PrimaryViewController(
        private val sourceCodeView: SourceCodeView,
        private val processTabPane: ProcessTabPane,
        private val fontSizeChanger: FontSizeChanger,
        private val themeManager: ThemeManager,
        private val fileChooser: Supplier<File>
) : Initializable {

    @FXML
    lateinit var primaryPane: VBox
    @FXML
    lateinit var processTabAnchorPane: AnchorPane
    @FXML
    lateinit var sourceCodeViewAnchorPane: AnchorPane

    override fun initialize(url: URL, rb: ResourceBundle?) {
        processTabAnchorPane.children.add(processTabPane)
        AnchorPane.setTopAnchor(processTabPane, 0.0)
        AnchorPane.setBottomAnchor(processTabPane, 0.0)
        AnchorPane.setRightAnchor(processTabPane, 0.0)
        AnchorPane.setLeftAnchor(processTabPane, 0.0)
        sourceCodeViewAnchorPane.children.add(sourceCodeView)
        AnchorPane.setTopAnchor(sourceCodeView, 0.0)
        AnchorPane.setBottomAnchor(sourceCodeView, 0.0)
        AnchorPane.setRightAnchor(sourceCodeView, 0.0)
        AnchorPane.setLeftAnchor(sourceCodeView, 0.0)
    }

    fun clearAll(event: Event?) {
        processTabPane.clear()
    }

    fun changeAggregation(event: Event?) {
        // TODO maybe implement
    }

    fun showAboutPopup(event: Event?) {
        val dialog = Stage()
        dialog.initModality(Modality.APPLICATION_MODAL)
        dialog.initOwner(Main.stage)
        val dialogVbox = AnchorPane()
        val text = StyledText.of("Ulyp 0.2, recording debugger", CssClass.TEXT)
        dialogVbox.children.add(text)

        AnchorPane.setTopAnchor(text, 50.0)
        AnchorPane.setBottomAnchor(text, 50.0)
        AnchorPane.setRightAnchor(text, 50.0)
        AnchorPane.setLeftAnchor(text, 50.0)

        val dialogScene = Scene(dialogVbox, 300.0, 200.0)
        dialogScene.stylesheets.add(Theme.DARK.ulypCssPath)
        dialog.scene = dialogScene
        dialog.show()
    }

    fun showControlsPopup(event: Event?) {

    }

    fun changeTheme(event: Event?) {
        themeManager.applyTheme(Theme.LIGHT, primaryPane.scene)
    }

    fun openRecordedDump(actionEvent: ActionEvent?) {
        // Without those calls font style won't be applied until user changes font for the first time
        fontSizeChanger.upscale(primaryPane.scene)
        fontSizeChanger.downscale(primaryPane.scene)
        val file = fileChooser.get()
        val storageReader = AsyncFileStorageReader(file, false)

        storageReader.processMetadataFuture.thenAccept { processMetadata ->
            storageReader.subscribe { recording ->
                val fileRecordingsTab = processTabPane.getOrCreateProcessTab(FileRecordingsTabName(file, processMetadata))
                val recordingTab = fileRecordingsTab.getOrCreateRecordingTab(processMetadata, recording)
                recordingTab.update(recording)
                Platform.runLater { recordingTab.refreshTreeView() }
            }
        }

        storageReader.start()
    }
}